subsys/net/lib/ocpp/core.c

```diff
@@ -0,0 +1,504 @@
+	FILL_METER_TABLE(OCPP_OMM_ACTIVE_POWER_TO_EV,
+			 "Power.Active.Import", "W"),
+	FILL_METER_TABLE(OCPP_OMM_REACTIVE_POWER_FROM_EV,
+			 "Power.Reactive.Export", "var"),
+	FILL_METER_TABLE(OCPP_OMM_REACTIVE_POWER_TO_EV,
+			 "Power.Reactive.Import", "var"),
+	FILL_METER_TABLE(OCPP_OMM_POWERLINE_FREQ,
+			 "Frequency", NULL),
+	FILL_METER_TABLE(OCPP_OMM_POWER_FACTOR,
+			 "Power.Factor", NULL),
+	FILL_METER_TABLE(OCPP_OMM_POWER_MAX_OFFERED_TO_EV,
+			 "Power.Offered", NULL),
+	FILL_METER_TABLE(OCPP_OMM_FAN_SPEED,
+			 "RPM", "rpm"),
+	FILL_METER_TABLE(OCPP_OMM_CHARGING_PERCENT,
+			 "SoCState", "Percent"),
+	FILL_METER_TABLE(OCPP_OMM_TEMPERATURE,
+			 "Temperature", "Celsius"),
+	FILL_METER_TABLE(OCPP_OMM_VOLTAGE_AC_RMS,
+			 "Voltage", "V")
+};
+
+int ocpp_boot_notification(ocpp_session_handle_t hndl,
+			   struct ocpp_cp_info *cpi)
+{
+	int ret;
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx = sh->ctx;
+	struct ocpp_upstream_info *ui = ctx->ui;
+	char *buf = ctx->pdu_buf;
+	ocpp_msg_fp_t fn;
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+
+	fn = ctx->cfn[PDU_BOOTNOTIFICATION];
+	sh->uid = fn(buf, sizeof(ctx->pdu_buf), sh, cpi);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	rmsg.sndlock = &ui->ws_sndlock;
+	rmsg.rspsig = &ui->ws_rspsig;
+
+	ret = ocpp_send_to_server(&rmsg, K_SECONDS(OCPP_PDU_TIMEOUT));
+	if (!ret && sh->resp_status != BOOT_ACCEPTED) {
+		ret = -EAGAIN;
+	}
+
+	return ret;
+}
+
+int ocpp_get_configuration(enum ocpp_key key, struct ocpp_info *ctx, char *uid)
+{
+	int ret;
+	char tmp[32];
+	char *sval = NULL;
+	char *buf = ctx->pdu_buf;
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	union ocpp_keyval *kval;
+	ocpp_msg_fp_t fn;
+	bool is_rw;
+	enum ocpp_key_type ktype;
+
+	if (key >= OCPP_CFG_END) {
+		return -EINVAL;
+	}
+
+	ktype = ocpp_get_keyval_type(key);
+	is_rw = ocpp_is_key_rw(key);
+	kval = ocpp_get_key_val(key);
+
+	if (ktype < KEY_TYPE_STR) {
+		sval = tmp;
+		snprintk(tmp, sizeof(tmp), "%d", kval->ival);
+	} else {
+		sval = kval->str;
+	}
+
+	fn = ctx->cfn[PDU_GET_CONFIGURATION];
+	fn(buf, sizeof(ctx->pdu_buf), ocpp_get_key_literal(key), sval, is_rw, uid);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+
+	ret = ocpp_send_to_server(&rmsg, K_SECONDS(OCPP_PDU_TIMEOUT));
+	return ret;
+}
+
+int ocpp_change_configuration(char *skey, struct ocpp_info *ctx,
+			      char *sval, char *uid)
+{
+	int ret = -EINVAL;
+	char *buf = ctx->pdu_buf;
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	union ocpp_keyval kval;
+	ocpp_msg_fp_t fn;
+	enum ocpp_key key;
+	char *res = "Accepted";
+
+	key = ocpp_key_to_cfg(skey);
+	if (key < OCPP_CFG_END) {
+		enum ocpp_key_type ktype;
+
+		ktype = ocpp_get_keyval_type(key);
+		if (ktype < KEY_TYPE_STR) {
+			kval.ival = atoi(sval);
+		} else {
+			kval.str = sval;
+		}
+		ret = ocpp_update_cfg_val(key, &kval);
+
+		if (ret) {
+			res = "Rejected";
+		}
+	} else {
+		res = "NotSupported";
+	}
+
+	if (!ret) {
```

Comment:
These was some rule (Misra?) that says we should only use boolean operator for boolean types. So here we should have `if (ret == 0) {` instead. This is not a big deal atm, but avoids future fixes if we done this right now.